File: leetcode_ly/java/List/PartitionList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode partition(ListNode head, int x) {
        if(head == null) return null;
        ListNode left = new ListNode(0);
        ListNode right = new ListNode(0);
        ListNode copyLeft = left;
        ListNode copyRight = right;
        while(head != null){
            ListNode node = head.next;
            if(head.val < x){
                left.next = head;
                head.next = null;
                left = left.next;
            }
            else{
                right.next = head;
                head.next = null;
                right = right.next;
            }
            head = node;
        }
        left.next = copyRight.next;
        return copyLeft.next;
    }
}

File: leetcode_ly/c++/List/PartitionList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        if(!head || !head->next) return head;
        ListNode* low = new ListNode(0);
        ListNode* lowCopy = low;
        ListNode* high = new ListNode(0);
        ListNode* highCopy = high;
        ListNode* index = head;
        while(index){
            ListNode* node = index->next;
            if(index->val < x){
                low->next = index;
                low = low->next;
                low->next = NULL;
            }
            else{
                high->next = index;
                high = high->next;
                high->next = NULL;
            }
            index = node;
        }
        low->next = highCopy->next;
        return lowCopy->next;
    }
};